js/apps/admin-ui/src/components/key-value-form/KeySelect.tsx

```diff
@@ -6,18 +6,40 @@ import { KeycloakTextInput } from "ui-shared";
 import useToggle from "../../utils/useToggle";
 import { DefaultValue } from "./KeyValueInput";
 
+/**
+ * Used when the values passed to KeySelect in "selectItems" need to be fetched from the server dynamically. In this case, we might want to call "setCustom"
+ * from parent component. The "fieldValue" sent to KeySelect is not supposed to be set by the calling component, as it will be set by KeySelect itself with the value of the field, so
+ * it can be consumed from the callback function once the values are fetched from the server
+ */
+export type FetchCallback = {
+  fieldValue: string;
+  custom: boolean;
+  setCustom: any;
+};
+
 type KeySelectProp = UseControllerProps & {
   selectItems: DefaultValue[];
+  fetchCallback?: FetchCallback;
 };
 
-export const KeySelect = ({ selectItems, ...rest }: KeySelectProp) => {
+export const KeySelect = ({
+  selectItems,
+  fetchCallback,
+  ...rest
+}: KeySelectProp) => {
   const { t } = useTranslation();
   const [open, toggle] = useToggle();
   const { field } = useController(rest);
-  const [custom, setCustom] = useState(
+
+  let [custom, setCustom] = useState(
     !selectItems.map(({ key }) => key).includes(field.value),
   );
 
+  if (fetchCallback) {
+    fetchCallback.fieldValue = field.value;
+    [custom, setCustom] = [fetchCallback.custom, fetchCallback.setCustom];
```

Comment:
It is not allowed to re-assign these values like this, these should be derived from that `useState` call at all times. These will be re-declared on each render either way.